src/ring_theory/integral_closure.lean

```diff
@@ -36,7 +36,12 @@ if it is a root of some monic polynomial `p : polynomial R`. -/
 def is_integral (x : A) : Prop :=
 ∃ p : polynomial R, monic p ∧ aeval x p = 0
 
-variable {R}
+variable (A)
+
+/-- An integral extension is one where every element of the -/
+def algebra.is_integral : Prop := ∀ x : A, is_integral R x
```

Comment:
I think it might be useful to have the predicate for `ring_hom`s as well. Do you think you can refactor the definitions above to work in that situation as well?

Reply:
It's already possible to define `def ring_hom.is_integral (f : R →+* S) : Prop := ∀ x : S, @is_integral R S _ _ f.to_algebra x`. Would you also want to refactor `is_integral` to operate on a ring hom instead of an algebra? Then all the existing uses would just pass in `(algebra_map R A)`, but I don't think `(algebra_map R A).to_algebra` is definitionally equal to the original algebra, so I'm worried that might not behave well.

Reply:
Or alternatively `def ring_hom.is_integral' {S : Type*} [comm_ring S] (f : R →+* S) : Prop := @algebra.is_integral R S _ _ f.to_algebra` could be the definition, which might be better so that theorems about `algebra.is_integral` also naturally apply to `ring_hom.is_integral`.

Reply:
My idea would be to introduce  refactor `is_integral` to operate on a ring hom, and then state the current definition for algebras in terms of that new definition.
Clearly, it is useful to make this stuff easy to use in the setting of algebras. But I guess we'll regret it if we don't provide the basic API in terms of ordinary ring homs.

Reply:
In particular, the "low-level" definitions shouldn't be stated in terms of the current definitions using algebras, but the other way round.

Reply:
I pushed changes to make the base definition `def is_integral (f : R →+* A) (x : A) := ∃ p : polynomial R, monic p ∧ eval₂ f x p = 0` and then the old definition is renamed to `def is_integral_alg (x : A) : Prop := is_integral (algebra_map R A) x`. Most of the changes then are just renaming all the old uses, but some proofs also needed to change if they relied on lemmas about `aeval`, since the definition is now in terms of `eval₂`.

Also I'm sure a lot of theorems can be generalized to assume a ring hom instead of an algebra, but I didn't make that change unless it was fairly easy to do (Like in `ring_theory/localization` where the algebra structure was somewhat artificial from the start. I can clean it up more but I wanted to get feedback on the general approach before going to far with these changes.